src/interpret.jl

```diff
@@ -157,11 +157,25 @@ function evaluate_foreigncall(frame::Frame, call_expr::Expr)
     scope = frame.framecode.scope
     data = frame.framedata
     if !isempty(data.sparams) && scope isa Method
-        sig = scope.sig
+        sig = scope.sig    
         args[2] = instantiate_type_in_env(args[2], sig, data.sparams)
-        args[3] = Core.svec(map(args[3]) do arg
-            instantiate_type_in_env(arg, sig, data.sparams)
-        end...)
+        @static if VERSION < v"1.7.0"
+            if args[3] isa Core.SimpleVector
+                args[3] = Core.svec(map(args[3]) do arg
+                    instantiate_type_in_env(arg, sig, data.sparams)
+                end...)
+            else
+                args[3] = instantiate_type_in_env(args[3], sig, data.sparams)
+                args[4] = Core.svec(map(args[4]) do arg
+                    instantiate_type_in_env(arg, sig, data.sparams)
+                end...)
+            end
+        else
+            args[3] = instantiate_type_in_env(args[3], sig, data.sparams)
+            args[4] = Core.svec(map(args[4]) do arg
+                instantiate_type_in_env(arg, sig, data.sparams)
+            end...)
+        end
```

Comment:
How much of the argument-position issue is specific to `:cfunction`? We may just be off by one index in `args` for `cfunction`.

Reply:
Will check.

Reply:
I searched for documentation about `:cfunction` in AST and found nothing ;( So i added some logging to test and see on 1.6.4

```
head = :foreigncall
args[1] = :(Core.tuple("cholmod_l_allocate_sparse", :libcholmod))
args[2] = Ptr{SuiteSparse.CHOLMOD.C_Sparse{Float64}}
args[3] = svec(UInt64, UInt64, UInt64, Int32, Int32, Int32, Int32, Ptr{Nothing})
head = :foreigncall
args[1] = :(Core.tuple("cholmod_l_check_sparse", :libcholmod))
args[2] = Int32
args[3] = svec(Ptr{SuiteSparse.CHOLMOD.C_Sparse{Tv<:Union{Float64, ComplexF64}}}, Ptr{UInt8})
head = :foreigncall
args[1] = :(Core.tuple("cholmod_l_analyze", :libcholmod))
args[2] = Ptr{SuiteSparse.CHOLMOD.C_Factor{Float64}}
args[3] = svec(Ptr{SuiteSparse.CHOLMOD.C_Sparse{Tv<:Union{Float64, ComplexF64}}}, Ptr{UInt8})
head = :foreigncall
args[1] = :(Core.tuple("cholmod_l_factorize_p", :libcholmod))
args[2] = Int32
args[3] = svec(Ptr{SuiteSparse.CHOLMOD.C_Sparse{Tv<:Union{Float64, ComplexF64}}}, Ref{ComplexF64}, Ptr{Int64}, UInt64, Ptr{SuiteSparse.CHOLMOD.C_Factor{Tv<:Union{Float64, ComplexF64}}}, Ptr{UInt8})
head = :foreigncall
args[1] = :(Core.tuple("cholmod_l_allocate_dense", :libcholmod))
args[2] = Ptr{SuiteSparse.CHOLMOD.C_Dense{Float64}}
args[3] = svec(UInt64, UInt64, UInt64, Int32, Ptr{Nothing})
head = :foreigncall
args[1] = :(Core.tuple("cholmod_l_solve", :libcholmod))
args[2] = Ptr{SuiteSparse.CHOLMOD.C_Dense{Float64}}
args[3] = svec(Int32, Ptr{SuiteSparse.CHOLMOD.C_Factor{Tv<:Union{Float64, ComplexF64}}}, Ptr{SuiteSparse.CHOLMOD.C_Dense{Tv<:Union{Float64, ComplexF64}}}, Ptr{UInt8})
head = :cfunction
args[1] = Ptr{Nothing}
args[2] = :($(QuoteNode(FunctionWrappers.CallWrapper{Int64}())))
args[3] = Int64
args[4] = svec(Ref{var"#35#36"})```
```

and on 1.7.0

```
head = :cfunction
args = Any[Ptr{Nothing}, :($(QuoteNode(FunctionWrappers.CallWrapper{Int64}()))), Int64, svec(Ref{var"#35#36"}), :(:ccall)]
```

Oops.

Reply:
More logging:

1.6.4:

```
head = :foreigncall
args[1] = :(:jl_sizeof_jl_options)
args[2] = Int64
args[3] = svec()
args[4] = 0
data.sparams = Any[]
scope = Base
head = :foreigncall
args[1] = :cf
args[2] = Int64
args[3] = svec(Int64, Int64)
args[4] = 0
data.sparams = Any[]
scope = Main
head = :foreigncall
args[1] = Ptr{Nothing} @0x000000005a8d5f90
args[2] = Int64
args[3] = svec(Int64, Int64)
args[4] = 0
data.sparams = Any[]
scope = Main
head = :cfunction
args[1] = Ptr{Nothing}
args[2] = :($(QuoteNode(fcfun)))
args[3] = Int64
args[4] = svec(Int64, Int64)
data.sparams = Any[]
scope = cfcfun() in Main at C:\Users\Win10\Documents\GitHub\JuliaInterpreter.jl\test\interpret.jl:162
head = :foreigncall
args[1] = Ptr{Nothing} @0x000000005a8d9630
args[2] = Int64
args[3] = svec(UInt8, Int64)
args[4] = 0
data.sparams = Any[]
scope = Main
head = :foreigncall
args[1] = :(Core.tuple("cholmod_l_allocate_sparse", :libcholmod))
args[2] = Ptr{SuiteSparse.CHOLMOD.C_Sparse{Tv<:Union{Float64, ComplexF64}}}
args[3] = svec(UInt64, UInt64, UInt64, Int32, Int32, Int32, Int32, Ptr{Nothing})
args[4] = 0
data.sparams = Any[Float64]
scope = allocate_sparse(nrow::Integer, ncol::Integer, nzmax::Integer, sorted::Bool, packed::Bool, stype::Integer, ::Type{Tv}) where Tv<:Union{Float64, ComplexF64} in SuiteSparse.CHOLMOD at C:\Users\Win10\AppData\Local\Programs\Julia-1.6.4\share\julia\stdlib\v1.6\SuiteSparse\src\cholmod.jl:483
head = :foreigncall
args[1] = :(Core.tuple("cholmod_l_check_sparse", :libcholmod))
args[2] = Int32
args[3] = svec(Ptr{SuiteSparse.CHOLMOD.C_Sparse{Tv<:Union{Float64, ComplexF64}}}, Ptr{UInt8})
args[4] = 0
data.sparams = Any[Float64]
scope = check_sparse(A::SuiteSparse.CHOLMOD.Sparse{Tv}) where Tv<:Union{Float64, ComplexF64} in SuiteSparse.CHOLMOD at C:\Users\Win10\AppData\Local\Programs\Julia-1.6.4\share\julia\stdlib\v1.6\SuiteSparse\src\cholmod.jl:542
head = :foreigncall
args[1] = :(Core.tuple("cholmod_l_analyze", :libcholmod))
args[2] = Ptr{SuiteSparse.CHOLMOD.C_Factor{Tv<:Union{Float64, ComplexF64}}}
args[3] = svec(Ptr{SuiteSparse.CHOLMOD.C_Sparse{Tv<:Union{Float64, ComplexF64}}}, Ptr{UInt8})
args[4] = 0
data.sparams = Any[Float64]
scope = analyze(A::SuiteSparse.CHOLMOD.Sparse{Tv}, cmmn::Vector{UInt8}) where Tv<:Union{Float64, ComplexF64} in SuiteSparse.CHOLMOD at C:\Users\Win10\AppData\Local\Programs\Julia-1.6.4\share\julia\stdlib\v1.6\SuiteSparse\src\cholmod.jl:718
head = :foreigncall
args[1] = :(Core.tuple("cholmod_l_factorize_p", :libcholmod))
args[2] = Int32
args[3] = svec(Ptr{SuiteSparse.CHOLMOD.C_Sparse{Tv<:Union{Float64, ComplexF64}}}, Ref{ComplexF64}, Ptr{Int64}, UInt64, Ptr{SuiteSparse.CHOLMOD.C_Factor{Tv<:Union{Float64, ComplexF64}}}, Ptr{UInt8})
args[4] = 0
data.sparams = Any[Float64]
scope = factorize_p!(A::SuiteSparse.CHOLMOD.Sparse{Tv}, β::Real, F::SuiteSparse.CHOLMOD.Factor{Tv}, cmmn::Vector{UInt8}) where Tv<:Union{Float64, ComplexF64} in SuiteSparse.CHOLMOD at C:\Users\Win10\AppData\Local\Programs\Julia-1.6.4\share\julia\stdlib\v1.6\SuiteSparse\src\cholmod.jl:739
head = :foreigncall
args[1] = :(Core.tuple("cholmod_l_allocate_dense", :libcholmod))
args[2] = Ptr{SuiteSparse.CHOLMOD.C_Dense{Tv<:Union{Float64, ComplexF64}}}
args[3] = svec(UInt64, UInt64, UInt64, Int32, Ptr{Nothing})
args[4] = 0
data.sparams = Any[Float64]
scope = allocate_dense(m::Integer, n::Integer, d::Integer, ::Type{Tv}) where Tv<:Union{Float64, ComplexF64} in SuiteSparse.CHOLMOD at C:\Users\Win10\AppData\Local\Programs\Julia-1.6.4\share\julia\stdlib\v1.6\SuiteSparse\src\cholmod.jl:413
head = :foreigncall
args[1] = :(Core.tuple("cholmod_l_solve", :libcholmod))
args[2] = Ptr{SuiteSparse.CHOLMOD.C_Dense{Tv<:Union{Float64, ComplexF64}}}
args[3] = svec(Int32, Ptr{SuiteSparse.CHOLMOD.C_Factor{Tv<:Union{Float64, ComplexF64}}}, Ptr{SuiteSparse.CHOLMOD.C_Dense{Tv<:Union{Float64, ComplexF64}}}, Ptr{UInt8})
args[4] = 0
data.sparams = Any[Float64]
scope = solve(sys::Integer, F::SuiteSparse.CHOLMOD.Factor{Tv}, B::SuiteSparse.CHOLMOD.Dense{Tv}) where Tv<:Union{Float64, ComplexF64} in SuiteSparse.CHOLMOD at C:\Users\Win10\AppData\Local\Programs\Julia-1.6.4\share\julia\stdlib\v1.6\SuiteSparse\src\cholmod.jl:749
head = :cfunction
args[1] = Ptr{Nothing}
args[2] = :($(QuoteNode(+)))
args[3] = Int64
args[4] = svec(Int64, Int64)
data.sparams = Any[]
scope = (::var"#f#34")() in Main at C:\Users\Win10\Documents\GitHub\JuliaInterpreter.jl\test\interpret.jl:786
head = :cfunction
args[1] = Ptr{Nothing}
args[2] = :($(QuoteNode(FunctionWrappers.CallWrapper{Int64}())))
args[3] = Int64
args[4] = svec(Ref{var"#35#36"})
data.sparams = Any[Int64, Tuple{}, var"#35#36"]
scope = gen_fptr(::Type{Ret}, ::Type{Args}, ::Type{objT}) where {Ret, Args, objT} in FunctionWrappers at C:\Users\Win10\.julia\packages\FunctionWrappers\8xdVB\src\FunctionWrappers.jl:81
head = :foreigncall
args[1] = :(:strlen)
args[2] = UInt64
args[3] = svec(Cstring)
args[4] = 0
data.sparams = Any[]
scope = Main.anonymous
head = :foreigncall
args[1] = :(:asprintf)
args[2] = Int32
args[3] = svec(Ptr{Ptr{Int8}}, Cstring, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, Float32, Float32, Float32, Float32, Float32, Float32, Float32, Float32, Float32)
args[4] = 2
data.sparams = Any[]
scope = Main.anonymous
head = :foreigncall
args[1] = :(:free)
args[2] = Nothing
args[3] = svec(Cstring)
args[4] = 0
data.sparams = Any[]
scope = Main.anonymous
```

On 1.7.0:

```
head = :foreigncall
args[1] = :(:jl_sizeof_jl_options)
args[2] = Int64
args[3] = svec()
args[4] = 0
data.sparams = Any[]
scope = Base
head = :foreigncall
args[1] = :cf
args[2] = Int64
args[3] = svec(Int64, Int64)
args[4] = 0
data.sparams = Any[]
scope = Main
head = :foreigncall
args[1] = Ptr{Nothing} @0x0000000062462e00
args[2] = Int64
args[3] = svec(Int64, Int64)
args[4] = 0
data.sparams = Any[]
scope = Main
head = :cfunction
args[1] = Ptr{Nothing}
args[2] = :($(QuoteNode(fcfun)))
args[3] = Int64
args[4] = svec(Int64, Int64)
data.sparams = Any[]
scope = cfcfun() in Main at C:\Users\Win10\Documents\GitHub\JuliaInterpreter.jl\test\interpret.jl:162
head = :foreigncall
args[1] = Ptr{Nothing} @0x00000000624665c0
args[2] = Int64
args[3] = svec(UInt8, Int64)
args[4] = 0
data.sparams = Any[]
scope = Main
head = :cfunction
args[1] = Ptr{Nothing}
args[2] = :($(QuoteNode(+)))
args[3] = Int64
args[4] = svec(Int64, Int64)
data.sparams = Any[]
scope = (::var"#f#34")() in Main at C:\Users\Win10\Documents\GitHub\JuliaInterpreter.jl\test\interpret.jl:786
head = :cfunction
args[1] = Ptr{Nothing}
args[2] = :($(QuoteNode(FunctionWrappers.CallWrapper{Int64}())))
args[3] = Int64
args[4] = svec(Ref{var"#35#36"})
data.sparams = Any[Int64, Tuple{}, var"#35#36"]
scope = gen_fptr(::Type{Ret}, ::Type{Args}, ::Type{objT}) where {Ret, Args, objT} in FunctionWrappers at C:\Users\Win10\.julia\packages\FunctionWrappers\8xdVB\src\FunctionWrappers.jl:81
head = :foreigncall
args[1] = :(:strlen)
args[2] = UInt64
args[3] = svec(Cstring)
args[4] = 0
data.sparams = Any[]
scope = Main.anonymous
head = :foreigncall
args[1] = :(:asprintf)
args[2] = Int32
args[3] = svec(Ptr{Ptr{Int8}}, Cstring, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, Float32, Float32, Float32, Float32, Float32, Float32, Float32, Float32, Float32)
args[4] = 2
data.sparams = Any[]
scope = Main.anonymous
head = :foreigncall
args[1] = :(:free)
args[2] = Nothing
args[3] = svec(Cstring)
args[4] = 0
data.sparams = Any[]
scope = Main.anonymous
```